File: services/server/src/safe_socket/safe_socket.py
import socket
import struct

HEADER_SIZE = 6



def recv_all(socket: socket.socket, size):
    bytesToRead = size
    bytesRead = 0
    chunks = []
    while bytesRead < bytesToRead:
        chunk = socket.recv(bytesToRead - bytesRead) 
        bytesRead += len(chunk)
        chunks.append(chunk)
        if chunk == b"":
            raise RuntimeError("socket connection broken")

    return b"".join(chunks)


def send_all(socket: socket.socket, bytes):
    bytesToWrite = len(bytes)
    bytesWritten = 0
    while bytesWritten < bytesToWrite:
        n = socket.send(bytes[bytesWritten:])
        bytesWritten += n
        if n == 0:
            raise RuntimeError("socket connection broken")
        
    return bytesWritten


def recv_frame(socket: socket.socket):
    header = recv_all(socket, HEADER_SIZE)
    opcode = header[0]
    agency = header[1]
    payload_size = struct.unpack(">I", header[2:])[0]
    payload = recv_all(socket, payload_size)
    return opcode, agency, payload


def send_frame(socket: socket.socket, opcode: int, agency: int, payload: bytes):
    header = bytes([opcode, agency]) + struct.pack(">I", len(payload))
    return send_all(socket, header + payload)

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import (
    "encoding/binary"
    "io"
)

const HEADER_SIZE = 6

const (
	OpSendBets     uint8 = 0x01
	OpEndBets     uint8 = 0x02
	OpWinnersList uint8 = 0x03
	OpAck         uint8 = 0x0A
	OpError       uint8 = 0x0B
)

//TODO: Complete with a short-read/short-write tolerant implementation

func SendAll(socket io.Writer, bytes []byte) error {
	bytesWritten := 0
	bytesToWrite := len(bytes)

	for bytesWritten < bytesToWrite {
		n, err := socket.Write(bytes[bytesWritten:])
		if err != nil {
			return err
		}
		bytesWritten += n
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	bytesRead := 0
	bytesToRead := size

	buff := make([]byte, size)
	for bytesRead < bytesToRead {

		n, err := socket.Read(buff[bytesRead:])
		bytesRead += n
		if err != nil {
			return buff[:bytesRead], err
		}
	}
	return buff, nil
}

func SendFrame(socket io.Writer, opcode byte, agency byte, payload []byte) error {
	header := make([]byte, HEADER_SIZE)
	header[0] = opcode
	header[1] = agency

	payloadSize := uint32(len(payload))
	binary.BigEndian.PutUint32(header[2:], payloadSize)

	if err := SendAll(socket, header); err != nil {
		return err
	}
	if len(payload) > 0 {
		return SendAll(socket, payload)
	}
	return nil
}

func RecvFrame(socket io.Reader) (byte, byte, []byte, error) {
	header, err := RecvAll(socket, HEADER_SIZE)
	if err != nil {
		return 0, 0, nil, err
	}

	opcode := header[0]
	agency := header[1]
	payloadSize := binary.BigEndian.Uint32(header[2:])
	payload, err := RecvAll(socket, int(payloadSize))
	if err != nil {
		return 0, 0, nil, err
	}

	return opcode, agency, payload, nil
}